File: android/src/main/java/com/supermap/RNUtils/LocationTransfer.java
package com.supermap.RNUtils;

import com.supermap.data.CoordSysTransMethod;
import com.supermap.data.CoordSysTransParameter;
import com.supermap.data.CoordSysTranslator;
import com.supermap.data.Point2D;
import com.supermap.data.Point2Ds;
import com.supermap.data.PrjCoordSys;
import com.supermap.data.PrjCoordSysType;
import com.supermap.interfaces.mapping.SMap;
import com.supermap.mapping.Map;
import com.supermap.mapping.MapControl;

/**
 * @Author: shanglongyang
 * Date:        2019/7/18
 * project:     iTablet
 * package:     iTablet
 * class:
 * description:
 */
public class LocationTransfer {
    public static Point2D mapCoordToLatitudeAndLongitude (Point2D point2D) {
        SMap sMap = SMap.getInstance();
        final MapControl mapControl = sMap.getSmMapWC().getMapControl();
        final Map map = mapControl.getMap();

        final Point2D pt = point2D;
        PrjCoordSys Prj = map.getPrjCoordSys();
        Point2Ds points = new Point2Ds();
        points.add(pt);
        PrjCoordSys desPrjCoorSys = new PrjCoordSys();
        desPrjCoorSys.setType(PrjCoordSysType.PCS_EARTH_LONGITUDE_LATITUDE);
        CoordSysTranslator.convert(points, Prj, desPrjCoorSys,
                new CoordSysTransParameter(),
                CoordSysTransMethod.MTH_GEOCENTRIC_TRANSLATION);

        pt.setX(points.getItem(0).getX());
        pt.setY(points.getItem(0).getY());
        return pt;
    }

    public static Point2Ds mapCoordToLatitudeAndLongitude (Point2Ds point2Ds) {
        SMap sMap = SMap.getInstance();
        final MapControl mapControl = sMap.getSmMapWC().getMapControl();
        final Map map = mapControl.getMap();

        PrjCoordSys Prj = map.getPrjCoordSys();
        PrjCoordSys desPrjCoorSys = new PrjCoordSys();
        desPrjCoorSys.setType(PrjCoordSysType.PCS_EARTH_LONGITUDE_LATITUDE);
        CoordSysTranslator.convert(point2Ds, Prj, desPrjCoorSys,
                new CoordSysTransParameter(),
                CoordSysTransMethod.MTH_GEOCENTRIC_TRANSLATION);
        return point2Ds;
    }

    public static Point2D latitudeAndLongitudeToMapCoord (Point2D point2D) {
        SMap sMap = SMap.getInstance();
        final MapControl mapControl = sMap.getSmMapWC().getMapControl();
        final Map map = mapControl.getMap();

        final Point2D pt = point2D;
        PrjCoordSys Prj = map.getPrjCoordSys();
        Point2Ds points = new Point2Ds();
        points.add(pt);
        PrjCoordSys desPrjCoorSys = new PrjCoordSys();
        desPrjCoorSys.setType(PrjCoordSysType.PCS_EARTH_LONGITUDE_LATITUDE);
        CoordSysTranslator.convert(points, desPrjCoorSys, Prj,
                new CoordSysTransParameter(),
                CoordSysTransMethod.MTH_GEOCENTRIC_TRANSLATION);

        pt.setX(points.getItem(0).getX());
        pt.setY(points.getItem(0).getY());
        return pt;
    }

    public static Point2Ds latitudeAndLongitudeToMapCoord (Point2Ds point2Ds) {
        SMap sMap = SMap.getInstance();
        final MapControl mapControl = sMap.getSmMapWC().getMapControl();
        final Map map = mapControl.getMap();

        PrjCoordSys Prj = map.getPrjCoordSys();
        PrjCoordSys desPrjCoorSys = new PrjCoordSys();
        desPrjCoorSys.setType(PrjCoordSysType.PCS_EARTH_LONGITUDE_LATITUDE);
        CoordSysTranslator.convert(point2Ds, desPrjCoorSys, Prj,
                new CoordSysTransParameter(),
                CoordSysTransMethod.MTH_GEOCENTRIC_TRANSLATION);
        return point2Ds;
    }
}

File: ios/native/utils/LocationTransfer.m
//
//  LocationTransfer.m
//  Supermap
//
//  Created by Shanglong Yang on 2019/7/18.
//  Copyright © 2019 Facebook. All rights reserved.
//

#import "LocationTransfer.h"

@implementation LocationTransfer
+ (Point2D *)mapCoordToLatitudeAndLongitudeByPoint:(Point2D *)point2D {
    Point2Ds *points = [[Point2Ds alloc]init];
    [points add:point2D];
    PrjCoordSys *srcPrjCoorSys = [[PrjCoordSys alloc]init];
    [srcPrjCoorSys setType:PCST_EARTH_LONGITUDE_LATITUDE];
    CoordSysTransParameter *param = [[CoordSysTransParameter alloc]init];
    
    [CoordSysTranslator convert:points PrjCoordSys:[[SMap singletonInstance].smMapWC.mapControl.map prjCoordSys] PrjCoordSys:srcPrjCoorSys CoordSysTransParameter:param CoordSysTransMethod:(CoordSysTransMethod)9603];
    point2D = [points getItem:0];
    
    return point2D;
}

+ (Point2Ds *)mapCoordToLatitudeAndLongitudeByPoints:(Point2Ds *)point2Ds {
    PrjCoordSys *srcPrjCoorSys = [[PrjCoordSys alloc]init];
    [srcPrjCoorSys setType:PCST_EARTH_LONGITUDE_LATITUDE];
    CoordSysTransParameter *param = [[CoordSysTransParameter alloc]init];
    
    [CoordSysTranslator convert:point2Ds PrjCoordSys:[[SMap singletonInstance].smMapWC.mapControl.map prjCoordSys] PrjCoordSys:srcPrjCoorSys CoordSysTransParameter:param CoordSysTransMethod:(CoordSysTransMethod)9603];
    
    return point2Ds;
}

+ (Point2D *)latitudeAndLongitudeToMapCoordByPoint:(Point2D *)point2D {
    Point2Ds *points = [[Point2Ds alloc]init];
    [points add:point2D];
    PrjCoordSys *srcPrjCoorSys = [[PrjCoordSys alloc]init];
    [srcPrjCoorSys setType:PCST_EARTH_LONGITUDE_LATITUDE];
    CoordSysTransParameter *param = [[CoordSysTransParameter alloc]init];
    
    [CoordSysTranslator convert:points PrjCoordSys:srcPrjCoorSys PrjCoordSys:[[SMap singletonInstance].smMapWC.mapControl.map prjCoordSys] CoordSysTransParameter:param CoordSysTransMethod:(CoordSysTransMethod)9603];
    point2D = [points getItem:0];

    return point2D;
}

+ (Point2Ds *)latitudeAndLongitudeToMapCoordByPoints:(Point2Ds *)point2Ds {
    PrjCoordSys *srcPrjCoorSys = [[PrjCoordSys alloc]init];
    [srcPrjCoorSys setType:PCST_EARTH_LONGITUDE_LATITUDE];
    CoordSysTransParameter *param = [[CoordSysTransParameter alloc]init];
    
    [CoordSysTranslator convert:point2Ds PrjCoordSys:srcPrjCoorSys PrjCoordSys:[[SMap singletonInstance].smMapWC.mapControl.map prjCoordSys] CoordSysTransParameter:param CoordSysTransMethod:(CoordSysTransMethod)9603];
    
    return point2Ds;
}
@end
